packages/react-instantsearch-core/src/connectors/connectInsights.ts

```diff
@@ -0,0 +1,77 @@
+import createConnector from '../core/createConnector';
+import { getResults } from '../core/indexUtils';
+
+type Without<T, K> = Pick<T, Exclude<keyof T, K>>;
+type Results = { index: string };
+type Hit = { objectID: string; __position: number; __queryID: string };
+
+type InsightsClient = (
+  method: InsightsClientMethod,
+  payload: InsightsClientPayload
+) => void;
+
+type InsightsClientMethod =
+  | 'clickedObjectIDsAfterSearch'
+  | 'convertedObjectIDsAfterSearch';
+
+type InsightsClientPayload = {
+  index: string;
+  queryID: string;
+  eventName: string;
+  objectIDs: string[];
+  positions?: number[];
+};
+
+function inferPayload({
+  method,
+  results,
+  currentHit,
+}: {
+  method: InsightsClientMethod;
+  results: Results;
+  currentHit: Hit;
+}):
+  | Without<InsightsClientPayload, 'eventName'>
+  | Without<InsightsClientPayload, 'eventName' | 'positions'> {
+  const { index } = results;
+  const queryID = currentHit.__queryID;
+  const objectIDs = [currentHit.objectID];
+  switch (method) {
+    case 'clickedObjectIDsAfterSearch': {
+      const positions = [currentHit.__position];
+      return { index, queryID, objectIDs, positions };
+    }
+
+    case 'convertedObjectIDsAfterSearch':
+      return { index, queryID, objectIDs };
+
+    default:
+      throw new Error(`Unsupported method passed ${method}`);
```

Comment:
Could use an error message with more context.

Reply:
Any suggestion on the wording for more context ?

Reply:
> Unknown method `"wrongMethod"` passed to the `insights` function.
> The supported methods are: `"clickedObjectIDsAfterSearch"`, `"convertedObjectIDsAfterSearch"`.
>
> See: https://algolia.com/doc/link-to-doc

Something along those lines.

We now have all the internal tools (TypeScript, documentation, etc.) to enforce those. Let's be as explicit as possible in our error messages. This doesn't really help our users.

Reply:
I agree. I fixed the message. 
I'll add the link to the docs in a later iteration (the docs are not up to date yet)